### DISCUSSION TITLE
How to keep dbt asset materialization for successfully materialized assets for failed runs?

### DISCUSSION THREAD

**Discussion by nixent** [dagster-io/dagster#17735] [Q&A] [answered]
Assume dbt asset consisting of two models `A->B`. If `A` is successful and `B` fails, whole run is registered as failed and `A` materialization is marked as failed as well. If `raise_on_error=False` then `A` will have materialization record, `B` - will not and whole run will be successful.  
As a workaround, it is possible to handle it by sending an alert on asset failure or have outdated asset notification based on freshness policy

 ```
@dbt_assets(manifest=Path("target", "manifest.json"))
def my_dbt_assets(context: AssetExecutionContext, dbt: DbtCliResource):
    dbt_run_invocation = dbt.cli(["run"], context=context, raise_on_error=False)

    if dbt_run_invocation.is_successful():
        yield from dbt_run_invocation.stream()
    else:
       send_asset_failure_alert(...)
```

I'm looking for a way to retain dbt asset materialization for successful assets and record run failure. In this case run status would be `partially successful`. This is also relevant for dbt tests with `severity=warn`

**Comment by rexledesma** [comment, marked as answer]
You could raise a [`Failure`](https://docs.dagster.io/concepts/ops-jobs-graphs/op-events#failures) in the body of your computation.

```python
@dbt_assets(manifest=Path("target", "manifest.json"))
def my_dbt_assets(context: AssetExecutionContext, dbt: DbtCliResource):
   dbt_run_invocation = dbt.cli(["run"], context=context, raise_on_error=False)

   yield from dbt_run_invocation.stream()
   
   if not dbt_run_invocation.is_successful():
       raise Failure(description="partial success")
```
